### PR TITLE
Use `coiled-runtime`

### PR DESCRIPTION
**Pull request by jrbourbeau**
This PR updates our default software environments to use the [`coiled-runtime` metapackage](https://github.com/coiled/coiled-runtime)

EDIT: cc @ncclementi @dchudz @ntabris @hayesgb

### CODE REVIEW

File: default/environment-py37.yml [hunk 1]
```diff
@@ -1,25 +1,7 @@
...
-  - pyarrow>=0.15.1
-  - python-graphviz
-  - scikit-learn>=0.23.1
-  - h5py
-  - xarray
-  - bottleneck
-  - requests
-  - pillow>=7.2.0
+  - coiled-runtime=0.0.3
+  - coiled=0.0.73
```
Comment (jrbourbeau) [inline default/environment-py37.yml:7]: Pinning to the latest `coiled` release explicitly to work around https://github.com/coiled/coiled-runtime/issues/22

**Comment by ntabris** [comment]
Not sure whether this is good or bad, but noting that at present coiled default envs have packages that aren't (explicitly) in the runtime metapackage, e.g., python-blosc, bokeh, h5py, xarray, pillow.

**Comment by jrbourbeau** [comment]
Yeah, that's a great point. Do we know if users / our examples rely on those packages? Like do we have an example where the default Coiled software environment is used for some Xarray computation? 

I also see that `coiled` supports Python 3.7 while `coiled-runtime` is `python >=3.8`. The pinned version of `dask`/`distributed` in `coiled-runtime` pre-date us dropping Python 3.7 support. So we could add it back in. @ntabris do you know `coiled`'s timeline is for dropping Python 3.7?

**Comment by jrbourbeau** [comment]
xref https://github.com/coiled/coiled-runtime/pull/47

**Comment by hayesgb** [comment]
I would suggest we allow users to install those packages (xarray, etc) alongside the coiled runtime, rather than adding them in, at least for the time being.  Thoughts?

**Comment by dchudz** [comment]
I don't know any timeline of dropping 3.7 support. Would want Todd to be involved in any discussion of that.

**Comment by ncclementi** [comment]
> I would suggest we allow users to install those packages (xarray, etc) alongside the coiled runtime, rather than adding them in, at least for the time being. Thoughts?

My only concern is that when installing `package-not-in-runtime` (e.g xarray) this package might have requirements that could bump the versions of what was installed with coiled-runtime. How we will handle support for that? Because it won't be an exact copy of what we offer.

**Comment by dchudz** [comment]
@mrocklin are we OK with updating the default env in a way that removes those packages that aren't in the coiled-runtime?

I imagine it's fine but wanted to check in for a product decision. This could break some users' stuff.

File: default/environment-py37.yml (hunk 1 above)
Comment (dchudz) [reply default/environment-py37.yml:7]: the software environment doesn't even care about the `coiled` package, does it?

I'm skeptical of this pin...

**Comment by mrocklin** [comment]
Yup.  Fine.  I'm curious, what are the packages?  Is this xarray and
friends?

**Comment by dchudz** [comment]
@mrocklin Nat gave us a (partial?) list above:

>  Not sure whether this is good or bad, but noting that at present coiled default envs have packages that aren't (explicitly) in the runtime metapackage, e.g., python-blosc, bokeh, h5py, xarray, pillow.

**Comment by jrbourbeau** [comment]
```python
In [42]: rt - d # packages in coiled-runtime, but not coiled-default
Out[42]:
{'click',
 'coiled',
 'distributed',
 'fsspec',
 'gcsfs',
 'ipywidgets',
 'jupyterlab',
 'labextension',
 'nb_conda_kernels',
 'nodejs',
 'snappy'}

In [43]: d - rt # packages in coiled-default, but not coiled-runtime
Out[43]:
{'blosc',
 'bokeh',
 'bottleneck',
 'cytoolz',
 'h5py',
 'image',
 'ml',
 'pillow',
 'requests',
 'xarray'}
```

**Comment by jrbourbeau** [comment]
So `xarray`, `dask-ml`, `dask-image`, `pillow`, `xarray`, and `h5py` are the relevant packages that won't be installed with this change. That seems okay to me.

**Comment by dchudz** [comment]
Seems okay to me too.

File: default/environment-py37.yml (hunk 1 above)
Comment (shughes-uk) [reply default/environment-py37.yml:7]: Yup. It's not used. Happy to remove it.

Comment (dchudz) [reply default/environment-py37.yml:7]: @jrbourbeau I'd approve this MR if you remove the coiled pin.

I know that means some solves will can an old coiled version, but that's fine, software environments don't care what coiled version we use.

**Comment by shughes-uk** [comment]
Can we validate against our examples? I know a lot of of them are broken and i'd like to at least note break more

**Comment by mrocklin** [comment]
> python-blosc, bokeh, h5py, xarray, pillow

We've dropped blosc.  Bokeh is included in the `dask` metapackage`.  h5py isn't that relevant for cloud.

xarray we should consider strongly.  pillow I'm fine waiting for a long while.  I'll raise a question about xarray in the runtime repository.  For now though please don't worry about this and just go ahead.

File: default/environment-py37.yml (hunk 1 above)
Comment (dchudz) [reply default/environment-py37.yml:7]: oh Samantha makes a good point, let's check our examples

**Comment by dchudz** [comment]
Are we already validating with the taxi example? I see this unit test... https://github.com/coiled/coiled-runtime/blob/main/tests/test_coiled.py

**Comment by mrocklin** [comment]
Issue for xarray https://github.com/coiled/coiled-runtime/issues/56

File: default/environment-py37.yml (hunk 1 above)
Comment (jrbourbeau) [reply default/environment-py37.yml:7]: Once https://github.com/coiled/coiled-runtime/pull/57 is in and released we can remove the `coiled` pin and users will get the latest `coiled` by default

**Comment by jrbourbeau** [comment]
I'd like to get the two `coiled-runtime` PRs in:

- [ ] https://github.com/coiled/coiled-runtime/pull/57
- [ ] https://github.com/coiled/coiled-runtime/pull/54

Then re-issue the `coiled-runtime` release and remove the `coiled=0.0.73` pin here. That should make everything green

**Comment by hayesgb** [comment]
@jrbourbeau @dchudz  -- Where did we land with Python 3.7 support?  It sounds like declarative requires 3.8.  Is this correct?

**Comment by dchudz** [comment]
no, as of this morning (`0.0.74`) v2 clusters (declarative) are OK on 3.7 too.

I guess you'd need to update the pin in this MR from 73 to 74. Or maybe the pin isn't needed anymore?

**Comment by jrbourbeau** [comment]
Yep, @dchudz is correct that Python 3.7 is good to go now and the `coiled` pin is no longer needed. My current plan of action is to:

1. Merge https://github.com/coiled/coiled-runtime/pull/62
2. Re-issue the `coiled-runtime=0.0.3` release (since we're still iterating quickly I'd prefer to minimize the number of `coiled-runtime` releases in the world)
3. Update this PR

**Comment by jrbourbeau** [comment]
Alright, I had to push a small fix to get the `coiled-runtime` repo to re-issue to the `coiled-runtime` release, but I think this should be all green now 🤞

**Comment by jrbourbeau** [comment]
> Can we validate against our examples? I know a lot of of them are broken and i'd like to at least note break more

I’ve not run code to confirm, but did look at the examples in https://docs.coiled.io/user_guide/examples.html and didn’t see anything that would lead me to believe we’re breaking anything that’s not already broken

**Comment by jrbourbeau** [comment]
Planning to merge this in an hour or so if no further comment

**Pull request merged**